docs/decisions/0011-depending-on-multiple-event-bus-implementations.rst

```diff
@@ -0,0 +1,80 @@
+
+We are not necessarily *satisfied* with this approach, but regard it as suitable for at least one release.
+
+Rejected Alternatives
+*********************
+
+There are several alternative approaches we have considered:
+
+- Have the implementation be part of the openedx-events package
+- Just have one implementation
+- Require all implementations to be installed out-of-tree
+- Depend on one default implementation in-tree, and require that others be installed
+
+Implement inside openedx-events
+===============================
+
+We briefly discussed the idea of including all of the implementations *in* the openedx-events package, possibly as extras (e.g. ``openedx-events[kafka]``). However, this would have complicated the development of openedx-events; it would have become impossible to e.g. pin the version of the Redis implementation in a dependent IDA while continuing to receive updates to the Kafka implementation. (The use of extras also would not have helped our dependencies issue; we'd then need to decide which extras to depend upon from IDAs' base requirements files.)
+
+Single implementation
+=====================
+
+It's very unlikely that one implementation would satisfy everyone. Any event store, queue, or message bus technology that is used is additional software that deployers must learn to install, manage, configure, and operate. The exceptions would be MySQL and Redis, which are already used in Open edX. However, neither of these was used for the first implementation. Therefore, even if Redis turns out to work very well, we'll still need the extension point.
+
+All out-of-tree
+===============
+
+As discussed, this results in more development friction (although possibly lower maintenance burden). The package also would not participate in our usual ``make upgrade`` process. This leads to either A) more operational risk if deployers leave the implementation package version unpinned in their deployment pipeline, or B) more toil if the version is pinned but has to be manually bumped in every IDA every time a new package version is released.
+
+Not having *any* implementation provided in a standard devstack installation also makes Event Bus work unnecessarily difficult.
+
+One default
+===========
+
+Another option would be to select one implementation to always keep in the dependencies, and require that others be installed out-of-tree. If we went this route, we would likely choose event-bus-redis as the default, as it does not require the deployer to run additional services. (Redis is already used in Open edX deployments.)
```

Comment:
I wonder if it makes sense to make a redis-as-default for development as a decision as part of this ADR? It wouldn't necessarily change what we are doing, but it would change how we write about it and how permanent this part of the decision might be. Even if we use a single ADR, maybe we make it more clear that deciding for kafka and redis are separate decisions, and we might be choosing to include them for different reasons (although potentially overlapping reasons as well).

Reply:
Added a bit about default-for-development.